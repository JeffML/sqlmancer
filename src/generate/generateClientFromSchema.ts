import _ from 'lodash'
import {
  isEnumType,
  isScalarType,
  GraphQLBoolean,
  GraphQLCompositeType,
  GraphQLEnumType,
  GraphQLID,
  GraphQLInt,
  GraphQLFloat,
  GraphQLList,
  GraphQLNonNull,
  GraphQLOutputType,
  GraphQLSchema,
  GraphQLString,
  GraphQLType,
} from 'graphql'
import { getArgumentValues } from 'graphql/execution/values'
import { GraphQLJSON, GraphQLJSONObject } from '../scalars'
import { Writable } from 'stream'

import {
  getModelDetails,
  getDirectiveArguments,
  getDirectiveByName,
  isCustomScalar,
  makeNullable,
  unwrap,
} from '../graphqlUtilities'
import { Dialect, FieldNameTransformation } from '../types'

const libraryPath = process.env.SQLMANCER_PATH || 'sqlmancer'

export function generateClientFromSchema(schema: GraphQLSchema, stream: Writable): void {
  const sqlmancerDirective = getDirectiveByName(schema.getQueryType(), 'sqlmancer')

  if (!sqlmancerDirective) {
    throw new Error(
      'Unable to parse Sqlmancer configuration from type definitions. Did you include the @sqlmancer directive on your Query type?'
    )
  }

  const options = getDirectiveArguments(sqlmancerDirective, schema)!
  const dialect = _.lowerCase(options.dialect) as Dialect
  const transformFieldNames = options.transformFieldNames as FieldNameTransformation

  stream.write(`
import Knex from 'knex'
import {
  BuilderOptions,
  CreateManyBuilder,
  CreateOneBuilder,
  DeleteByIdBuilder,
  DeleteManyBuilder,
  FindByIdBuilder,
  FindManyBuilder,
  FindOneBuilder,
  Models,
  UpdateByIdBuilder,
  UpdateManyBuilder
} from '${libraryPath}'

export type JSON = boolean | number | string | null | JSONArray | JSONMap
export interface JSONMap {
  [key: string]: JSON
}
export type JSONArray = Array<JSON>

export const models: Models = {`)

  const models = getModels(schema).map(model => {
    const modelDirective = getDirectiveByName(model, 'model')!
    const { table, pk, include } = getArgumentValues(schema.getDirective('model')!, modelDirective)
    const details = getModelDetails(model, schema)
    return {
      ...model,
      details,
      tableName: table,
      primaryKey: pk,
      include,
    }
  })

  models.forEach(({ name, tableName, primaryKey, include, details: { fields, joins, dependencies } }) => {
    stream.write(
      `
  ${name}: {
    tableName: '${tableName}',
    primaryKey: '${primaryKey}',
    fields: {\n${fields
      .map(
        ({ fieldName, columnName, hasDefault }) => `      ${fieldName}: {
        column: '${columnName || transformFieldName(fieldName, transformFieldNames)}',${
          hasDefault ? '\n        hasDefault: true,\n' : ''
        }
      },`
      )
      .join('\n')}
    },
    include: ${JSON.stringify(include || [])},
    dependencies: {\n${dependencies
      .map(({ fieldName, columns }) => `      ${fieldName}: [${columns.map(col => `'${col}'`).join(', ')}],`)
      .join('\n')}
    },
    associations: {\n${joins
      .map(({ fieldName, type, on, through }) => {
        const isList = makeNullable(type) instanceof GraphQLList
        const associationType = unwrap(type)

        const isModel = !!models.find(model => model.name === associationType.name)
        if (!isModel) {
          throw new Error(
            `The @join directive was used on ${name}.${fieldName} but the type ${associationType.name} is not a model.`
          )
        }

        return `      ${fieldName}: {
        modelName: '${associationType.name}',
        isMany: ${isList},
        on: [\n${on
          .map(
            ({ from, to }) => `          {
            from: '${from}',
            to: '${to}',
          },`
          )
          .join('\n')}
        ],
        through: ${through ? `'${through}'` : 'undefined'},
        builder: options => new ${associationType.name}Find${isList ? 'Many' : 'One'}Builder(options),
      },`
      })
      .join('\n')}
    },
  },`
    )
  })

  stream.write('\n}\n')

  models.forEach(({ name, primaryKey, details: { fields, joins } }) => {
    const enums: Record<string, GraphQLEnumType> = {}

    stream.write(`
export type ${name}Fields = {\n${fields
      .map(({ fieldName, type }) => `  ${getFieldProperty(fieldName, type, enums)}`)
      .join('\n')}
}
    `)

    const idFields = fields.filter(({ type }) => unwrap(type) === GraphQLID)
    stream.write(`
export type ${name}Ids = ${idFields.length ? idFields.map(({ fieldName }) => `'${fieldName}'`).join(' | ') : 'unknown'}
    `)

    const enumNames = Object.keys(enums)
    stream.write(`
export type ${name}Enums = ${enumNames.length ? enumNames.join(' | ') : 'unknown'}
    `)

    stream.write(`
export type ${name}Associations = {\n${joins
      .map(({ fieldName, type }) => `  ${getAssociationProperty(fieldName, type)}`)
      .join('\n')}
}
    `)

    stream.write(`
export type ${name}CreateFields = {\n${fields
      .map(
        ({ fieldName, type, hasDefault }) =>
          `  ${getFieldProperty(fieldName, hasDefault ? makeNullable(type) : type, enums)}`
      )
      .join('\n')}
}
    `)

    stream.write(`
export type ${name}UpdateFields = {\n${fields
      .filter(({ fieldName }) => fieldName !== primaryKey)
      .map(({ fieldName, type }) => `  ${getFieldProperty(fieldName, makeNullable(type), enums)}`)
      .join('\n')}
}
    `)

    Object.keys(enums).forEach(enumName => {
      const enumType = enums[enumName]
      stream.write(`
enum ${enumName} {\n${enumType
        .getValues()
        .map(enumValue => `  ${enumValue.name} = ${JSON.stringify(enumValue.value)},`)
        .join('\n')}
}`)
    })

    stream.write(
      `
export class ${name}FindOneBuilder<TSelected extends Pick<${name}Fields, any> = ${name}Fields> extends FindOneBuilder<
  '${dialect}',
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations,
  TSelected
> {
  constructor(options: BuilderOptions) {
    super(options, '${name}', models)
  }
}

export class ${name}FindManyBuilder<TSelected extends Pick<${name}Fields, any> = ${name}Fields> extends FindManyBuilder<
  '${dialect}',
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations,
  TSelected
> {
  constructor(options: BuilderOptions) {
    super(options, '${name}', models)
  }
}

export class ${name}FindByIdBuilder<TSelected extends Pick<${name}Fields, any> = ${name}Fields> extends FindByIdBuilder<
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations,
  TSelected
> {
  constructor(options: BuilderOptions, pk: number | string) {
    super(options, '${name}', models, pk)
  }
}
      `
    )

    stream.write(`
export class ${name}DeleteManyBuilder extends DeleteManyBuilder<
  '${dialect}',
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations
> {
  constructor(options: BuilderOptions) {
    super(options, '${name}', models)
  }
}

export class ${name}DeleteByIdBuilder extends DeleteByIdBuilder<${name}Fields, ${name}Ids, ${name}Enums, ${name}Associations> {
  constructor(options: BuilderOptions, pk: number | string) {
    super(options, '${name}', models, pk)
  }
}

export class ${name}CreateManyBuilder extends CreateManyBuilder<${name}CreateFields> {
  constructor(options: BuilderOptions, data: ${name}CreateFields[]) {
    super(options, '${name}', models, data)
  }
}

export class ${name}CreateOneBuilder extends CreateOneBuilder<${name}CreateFields> {
  constructor(options: BuilderOptions, data: ${name}CreateFields) {
    super(options, '${name}', models, data)
  }
}

export class ${name}UpdateManyBuilder extends UpdateManyBuilder<
  '${dialect}',
  ${name}UpdateFields,
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations
> {
  constructor(options: BuilderOptions, data: ${name}UpdateFields) {
    super(options, '${name}', models, data)
  }
}

export class ${name}UpdateByIdBuilder extends UpdateByIdBuilder<${name}UpdateFields> {
  constructor(options: BuilderOptions, pk: number | string, data: ${name}UpdateFields) {
    super(options, '${name}', models, pk, data)
  }
}
  `)
  })

  stream.write(`
type SqlmancerClient = {
  raw: Knex.RawBuilder<any>;
  transactionProvider(
    config?: any
  ): () => Promise<Knex.Transaction>;
  transaction(
    transactionScope?: null,
    config?: any
  ): Promise<Knex.Transaction>;
  transaction<T>(
    transactionScope: (trx: Knex.Transaction) => Promise<T> | Promise<T> | void,
    config?: any
  ): Promise<T>;
  initialize(config?: Knex.Config): void;
  destroy(callback: Function): void;
  destroy(): Promise<void>;
  batchInsert(
    tableName: Knex.TableDescriptor,
    data: readonly any[],
    chunkSize?: number
  ): Knex.QueryBuilder<any, {}>;
  schema: Knex.SchemaBuilder;
  queryBuilder<TRecord2 = any, TResult2 = any[]>(): Knex.QueryBuilder<TRecord2, TResult2>;
  client: any;
  migrate: Knex.Migrator;
  seed: Knex.Seeder;
  fn: Knex.FunctionHelper;
  ref: Knex.RefBuilder;
  models: {${models
    .map(
      ({ name }) => `
    ${name}: {
      findById: (id: number | string) => ${name}FindByIdBuilder
      findMany: () => ${name}FindManyBuilder
      findOne: () => ${name}FindOneBuilder
      createMany: (data: ${name}CreateFields[]) => ${name}CreateManyBuilder
      createOne: (data: ${name}CreateFields) => ${name}CreateOneBuilder
      deleteById: (id: number | string) => ${name}DeleteByIdBuilder
      deleteMany: () => ${name}DeleteManyBuilder
      updateById: (id: number | string, data: ${name}UpdateFields) => ${name}UpdateByIdBuilder
      updateMany: (data: ${name}UpdateFields) => ${name}UpdateManyBuilder
    },`
    )
    .join('')}
  },
}

export function createClient (knex: Knex): SqlmancerClient {
  return {
    models: {${models
      .map(
        ({ name }) => `
      ${name}: {
        findById: (id: number | string) => new ${name}FindByIdBuilder({ knex, dialect: '${dialect}' }, id),
        findMany: () => new ${name}FindManyBuilder({ knex, dialect: '${dialect}' }),
        findOne: () => new ${name}FindOneBuilder({ knex, dialect: '${dialect}' }),
        createMany: (data: ${name}CreateFields[]) => new ${name}CreateManyBuilder({ knex, dialect: '${dialect}' }, data),
        createOne: (data: ${name}CreateFields) => new ${name}CreateOneBuilder({ knex, dialect: '${dialect}' }, data),
        deleteById: (id: number | string) => new ${name}DeleteByIdBuilder({ knex, dialect: '${dialect}' }, id),
        deleteMany: () => new ${name}DeleteManyBuilder({ knex, dialect: '${dialect}' }),
        updateById: (id: number | string, data: ${name}UpdateFields) => new ${name}UpdateByIdBuilder({ knex, dialect: '${dialect}' }, id, data),
        updateMany: (data: ${name}UpdateFields) => new ${name}UpdateManyBuilder({ knex, dialect: '${dialect}' }, data),
      },`
      )
      .join('')}
    },
    ...knex,
  }
}
  `)
}

function transformFieldName(fieldName: string, transformation?: FieldNameTransformation) {
  switch (transformation) {
    case 'CAMEL_CASE':
      return _.camelCase(fieldName)
    case 'PASCAL_CASE':
      return _.upperFirst(_.camelCase(fieldName))
    case 'SNAKE_CASE':
      return _.snakeCase(fieldName)
    default:
      return fieldName
  }
}

function getFieldProperty(fieldName: string, type: GraphQLType, enums: Record<string, GraphQLEnumType>): string {
  const nullableType = makeNullable(type)
  const unwrappedType = unwrap(type)
  const colon = type instanceof GraphQLNonNull ? ': ' : '?: '
  const brackets = nullableType instanceof GraphQLList ? '[]' : ''
  if (isScalarType(unwrappedType)) {
    let tsType

    if (unwrappedType === GraphQLString) {
      tsType = 'string'
    } else if (unwrappedType === GraphQLInt || unwrappedType === GraphQLFloat) {
      tsType = 'number'
    } else if (unwrappedType === GraphQLBoolean) {
      tsType = 'boolean'
    } else if (unwrappedType === GraphQLID) {
      tsType = '(number | string)'
    } else if (isCustomScalar(unwrappedType, GraphQLJSON) || isCustomScalar(unwrappedType, GraphQLJSONObject)) {
      tsType = 'JSON'
    }
    return tsType ? `${fieldName}${colon}${tsType}${brackets}` : ''
  } else if (isEnumType(unwrappedType)) {
    enums[unwrappedType.name] = unwrappedType
    return `${fieldName}${colon}${unwrappedType.name}${brackets}`
  }
  return ''
}

function getAssociationProperty(associationName: string, type: GraphQLOutputType): string {
  const nullableType = makeNullable(type)
  const isList = nullableType instanceof GraphQLList
  const unwrappedType = unwrap(type)
  return `${associationName}: ${unwrappedType.name}Find${isList ? 'Many' : 'One'}Builder`
}

function getModels(schema: GraphQLSchema) {
  return Object.keys(schema.getTypeMap()).reduce((acc, typeName) => {
    const type = schema.getType(typeName)!
    const modelDirective = getDirectiveByName(type, 'model')
    if (modelDirective) {
      acc.push(type as GraphQLCompositeType)
    }
    return acc
  }, [] as GraphQLCompositeType[])
}
