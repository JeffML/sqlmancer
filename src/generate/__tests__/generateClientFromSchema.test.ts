import { WritableMock } from 'stream-mock'
import { generateClientFromSchema } from '../generateClientFromSchema'
import { schema } from './__fixtures__/schema'

describe('generateClientFromSchema', () => {
  test('correct usage', async () => {
    const stream = new WritableMock()
    generateClientFromSchema(schema, stream)
    stream.end()
    await new Promise((resolve, reject) => {
      stream.on('finish', () => {
        expect((stream.flatData as Buffer).toString()).toMatchSnapshot()
        resolve()
      })
      stream.on('error', reject)
    })
  })
})
