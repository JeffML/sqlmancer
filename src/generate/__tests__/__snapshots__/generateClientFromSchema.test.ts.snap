// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateClientFromSchema correct usage 1`] = `
"
import Knex from 'knex'
import {
  BuilderOptions,
  CreateManyBuilder,
  CreateOneBuilder,
  DeleteByIdBuilder,
  DeleteManyBuilder,
  FindByIdBuilder,
  FindManyBuilder,
  FindOneBuilder,
  Models,
  UpdateByIdBuilder,
  UpdateManyBuilder
} from 'sqlmancer'

export type JSON = boolean | number | string | null | JSONArray | JSONMap
export interface JSONMap {
  [key: string]: JSON
}
export type JSONArray = Array<JSON>

export const models: Models = {
  Actor: {
    tableName: 'actor',
    primaryKey: 'actor_id',
    fields: {
      id: {
        column: 'actor_id',
      },
      firstName: {
        column: 'first_name',
      },
      lastName: {
        column: 'last_name',
      },
    },
    include: [],
    dependencies: {

    },
    associations: {
      films: {
        modelName: 'Film',
        isMany: true,
        on: [
          {
            from: 'actor_id',
            to: 'actor_id',
          },
          {
            from: 'film_id',
            to: 'film_id',
          },
        ],
        through: 'film_actor',
        builder: options => new FilmFindManyBuilder(options),
      },
    },
  },
  Film: {
    tableName: 'film',
    primaryKey: 'film_id',
    fields: {
      id: {
        column: 'film_id',
      },
      title: {
        column: 'title',
      },
      releaseYear: {
        column: 'release_year',
      },
    },
    include: [],
    dependencies: {

    },
    associations: {
      actors: {
        modelName: 'Actor',
        isMany: true,
        on: [
          {
            from: 'film_id',
            to: 'film_id',
          },
          {
            from: 'actor_id',
            to: 'actor_id',
          },
        ],
        through: 'film_actor',
        builder: options => new ActorFindManyBuilder(options),
      },
    },
  },
}

export type ActorFields = {
  id: (number | string)
  firstName: string
  lastName: string
}
    
export type ActorIds = 'id'
    
export type ActorEnums = unknown
    
export type ActorAssociations = {
  films: FilmFindManyBuilder
}
    
export type ActorCreateFields = {
  id: (number | string)
  firstName: string
  lastName: string
}
    
export type ActorUpdateFields = {
  id?: (number | string)
  firstName?: string
  lastName?: string
}
    
export class ActorFindOneBuilder<TSelected extends Pick<ActorFields, any> = ActorFields> extends FindOneBuilder<
  ActorFields,
  ActorIds,
  ActorEnums,
  ActorAssociations,
  TSelected
> {
  constructor(options: BuilderOptions) {
    super(options, 'Actor', models)
  }
}

export class ActorFindManyBuilder<TSelected extends Pick<ActorFields, any> = ActorFields> extends FindManyBuilder<
  ActorFields,
  ActorIds,
  ActorEnums,
  ActorAssociations,
  TSelected
> {
  constructor(options: BuilderOptions) {
    super(options, 'Actor', models)
  }
}

export class ActorFindByIdBuilder<TSelected extends Pick<ActorFields, any> = ActorFields> extends FindByIdBuilder<
  ActorFields,
  ActorIds,
  ActorEnums,
  ActorAssociations,
  TSelected
> {
  constructor(options: BuilderOptions, pk: number | string) {
    super(options, 'Actor', models, pk)
  }
}
      
export class ActorDeleteManyBuilder extends DeleteManyBuilder<ActorFields, ActorIds, ActorEnums, ActorAssociations> {
  constructor(options: BuilderOptions) {
    super(options, 'Actor', models)
  }
}

export class ActorDeleteByIdBuilder extends DeleteByIdBuilder<ActorFields, ActorIds, ActorEnums, ActorAssociations> {
  constructor(options: BuilderOptions, pk: number | string) {
    super(options, 'Actor', models, pk)
  }
}

export class ActorCreateManyBuilder extends CreateManyBuilder<ActorCreateFields> {
  constructor(options: BuilderOptions, data: ActorCreateFields[]) {
    super(options, 'Actor', models, data)
  }
}

export class ActorCreateOneBuilder extends CreateOneBuilder<ActorCreateFields> {
  constructor(options: BuilderOptions, data: ActorCreateFields) {
    super(options, 'Actor', models, data)
  }
}

export class ActorUpdateManyBuilder extends UpdateManyBuilder<
  ActorUpdateFields,
  ActorFields,
  ActorIds,
  ActorEnums,
  ActorAssociations
> {
  constructor(options: BuilderOptions, data: ActorUpdateFields) {
    super(options, 'Actor', models, data)
  }
}

export class ActorUpdateByIdBuilder extends UpdateByIdBuilder<ActorUpdateFields> {
  constructor(options: BuilderOptions, pk: number | string, data: ActorUpdateFields) {
    super(options, 'Actor', models, pk, data)
  }
}
  
export type FilmFields = {
  id: (number | string)
  title: string
  releaseYear: number
}
    
export type FilmIds = 'id'
    
export type FilmEnums = unknown
    
export type FilmAssociations = {
  actors: ActorFindManyBuilder
}
    
export type FilmCreateFields = {
  id: (number | string)
  title: string
  releaseYear: number
}
    
export type FilmUpdateFields = {
  id?: (number | string)
  title?: string
  releaseYear?: number
}
    
export class FilmFindOneBuilder<TSelected extends Pick<FilmFields, any> = FilmFields> extends FindOneBuilder<
  FilmFields,
  FilmIds,
  FilmEnums,
  FilmAssociations,
  TSelected
> {
  constructor(options: BuilderOptions) {
    super(options, 'Film', models)
  }
}

export class FilmFindManyBuilder<TSelected extends Pick<FilmFields, any> = FilmFields> extends FindManyBuilder<
  FilmFields,
  FilmIds,
  FilmEnums,
  FilmAssociations,
  TSelected
> {
  constructor(options: BuilderOptions) {
    super(options, 'Film', models)
  }
}

export class FilmFindByIdBuilder<TSelected extends Pick<FilmFields, any> = FilmFields> extends FindByIdBuilder<
  FilmFields,
  FilmIds,
  FilmEnums,
  FilmAssociations,
  TSelected
> {
  constructor(options: BuilderOptions, pk: number | string) {
    super(options, 'Film', models, pk)
  }
}
      
export class FilmDeleteManyBuilder extends DeleteManyBuilder<FilmFields, FilmIds, FilmEnums, FilmAssociations> {
  constructor(options: BuilderOptions) {
    super(options, 'Film', models)
  }
}

export class FilmDeleteByIdBuilder extends DeleteByIdBuilder<FilmFields, FilmIds, FilmEnums, FilmAssociations> {
  constructor(options: BuilderOptions, pk: number | string) {
    super(options, 'Film', models, pk)
  }
}

export class FilmCreateManyBuilder extends CreateManyBuilder<FilmCreateFields> {
  constructor(options: BuilderOptions, data: FilmCreateFields[]) {
    super(options, 'Film', models, data)
  }
}

export class FilmCreateOneBuilder extends CreateOneBuilder<FilmCreateFields> {
  constructor(options: BuilderOptions, data: FilmCreateFields) {
    super(options, 'Film', models, data)
  }
}

export class FilmUpdateManyBuilder extends UpdateManyBuilder<
  FilmUpdateFields,
  FilmFields,
  FilmIds,
  FilmEnums,
  FilmAssociations
> {
  constructor(options: BuilderOptions, data: FilmUpdateFields) {
    super(options, 'Film', models, data)
  }
}

export class FilmUpdateByIdBuilder extends UpdateByIdBuilder<FilmUpdateFields> {
  constructor(options: BuilderOptions, pk: number | string, data: FilmUpdateFields) {
    super(options, 'Film', models, pk, data)
  }
}
  
type SqlmancerClient = {
  raw: Knex.RawBuilder<any>;
  transactionProvider(
    config?: any
  ): () => Promise<Knex.Transaction>;
  transaction(
    transactionScope?: null,
    config?: any
  ): Promise<Knex.Transaction>;
  transaction<T>(
    transactionScope: (trx: Knex.Transaction) => Promise<T> | Promise<T> | void,
    config?: any
  ): Promise<T>;
  initialize(config?: Knex.Config): void;
  destroy(callback: Function): void;
  destroy(): Promise<void>;
  batchInsert(
    tableName: Knex.TableDescriptor,
    data: readonly any[],
    chunkSize?: number
  ): Knex.QueryBuilder<any, {}>;
  schema: Knex.SchemaBuilder;
  queryBuilder<TRecord2 = any, TResult2 = any[]>(): Knex.QueryBuilder<TRecord2, TResult2>;
  client: any;
  migrate: Knex.Migrator;
  seed: Knex.Seeder;
  fn: Knex.FunctionHelper;
  ref: Knex.RefBuilder;
  models: {
    Actor: {
      findById: (id: number | string) => ActorFindByIdBuilder
      findMany: () => ActorFindManyBuilder
      findOne: () => ActorFindOneBuilder
      createMany: (data: ActorCreateFields[]) => ActorCreateManyBuilder
      createOne: (data: ActorCreateFields) => ActorCreateOneBuilder
      deleteById: (id: number | string) => ActorDeleteByIdBuilder
      deleteMany: () => ActorDeleteManyBuilder
      updateById: (id: number | string, data: ActorUpdateFields) => ActorUpdateByIdBuilder
      updateMany: (data: ActorUpdateFields) => ActorUpdateManyBuilder
    },
    Film: {
      findById: (id: number | string) => FilmFindByIdBuilder
      findMany: () => FilmFindManyBuilder
      findOne: () => FilmFindOneBuilder
      createMany: (data: FilmCreateFields[]) => FilmCreateManyBuilder
      createOne: (data: FilmCreateFields) => FilmCreateOneBuilder
      deleteById: (id: number | string) => FilmDeleteByIdBuilder
      deleteMany: () => FilmDeleteManyBuilder
      updateById: (id: number | string, data: FilmUpdateFields) => FilmUpdateByIdBuilder
      updateMany: (data: FilmUpdateFields) => FilmUpdateManyBuilder
    },
  },
}

export function createClient (knex: Knex): SqlmancerClient {
  return {
    models: {
      Actor: {
        findById: (id: number | string) => new ActorFindByIdBuilder({ knex, dialect: 'postgres' }, id),
        findMany: () => new ActorFindManyBuilder({ knex, dialect: 'postgres' }),
        findOne: () => new ActorFindOneBuilder({ knex, dialect: 'postgres' }),
        createMany: (data: ActorCreateFields[]) => new ActorCreateManyBuilder({ knex, dialect: 'postgres' }, data),
        createOne: (data: ActorCreateFields) => new ActorCreateOneBuilder({ knex, dialect: 'postgres' }, data),
        deleteById: (id: number | string) => new ActorDeleteByIdBuilder({ knex, dialect: 'postgres' }, id),
        deleteMany: () => new ActorDeleteManyBuilder({ knex, dialect: 'postgres' }),
        updateById: (id: number | string, data: ActorUpdateFields) => new ActorUpdateByIdBuilder({ knex, dialect: 'postgres' }, id, data),
        updateMany: (data: ActorUpdateFields) => new ActorUpdateManyBuilder({ knex, dialect: 'postgres' }, data),
      },
      Film: {
        findById: (id: number | string) => new FilmFindByIdBuilder({ knex, dialect: 'postgres' }, id),
        findMany: () => new FilmFindManyBuilder({ knex, dialect: 'postgres' }),
        findOne: () => new FilmFindOneBuilder({ knex, dialect: 'postgres' }),
        createMany: (data: FilmCreateFields[]) => new FilmCreateManyBuilder({ knex, dialect: 'postgres' }, data),
        createOne: (data: FilmCreateFields) => new FilmCreateOneBuilder({ knex, dialect: 'postgres' }, data),
        deleteById: (id: number | string) => new FilmDeleteByIdBuilder({ knex, dialect: 'postgres' }, id),
        deleteMany: () => new FilmDeleteManyBuilder({ knex, dialect: 'postgres' }),
        updateById: (id: number | string, data: FilmUpdateFields) => new FilmUpdateByIdBuilder({ knex, dialect: 'postgres' }, id, data),
        updateMany: (data: FilmUpdateFields) => new FilmUpdateManyBuilder({ knex, dialect: 'postgres' }, data),
      },
    },
    ...knex,
  }
}
  "
`;
